README.md
# ph5
C++ library for Pythagorean Hodograph Quintic curves

### Overview
Pythogarean Hodograph curves provide a clean and elegant representation for
machine traversable curves without the control point clutter that plagues cubic splines.
Quintic PH curves provide C2 continuity ideal for smooth machine acceleration.

This library is targeted for use in FirePick Delta machines, but may find use
elsewhere.

### What is a PH5Curve?
A Pythogarean Hodograph curve is a polynomial approximation to 2D curve. As such,
it is similar to a Bezier or NURBS curve. What makes a PH curve special is that
the curve passes through its control points. The on-curve control point greatly
reduces the human labor of specifying a PH curve (In reality, a PH curve does in
fact have off-curve control points, but these are secondary and entirely derived
from the on-curve control points.)

PH curves, in particular, quintic PH curves (PH5Curve) are well suited to generating
smooth curves for CNC robots. Although higher order PH curves (7,9, etc.) are also
acceptable, the PH5Curve exhibits C2 continuity for the least computational burden.
C2 continuity is important because C2 continuity asserts that the acceleration profile for a
point traversing the curve is continuous. If acceleration is discontinuous,
machinery jerks.

### Computation and Memory
Generating PH curves from a set of control points is computationally and memory 
intensive. However, traversing a PH curve is less demanding. And, finally,
traversing a piece-wise linear approximation to a a sampled PH curve is 
even less demanding.

This library is designed to run on processors as small as the ArduinoMega2560,
which only has 8KB of free RAM. On sucha a small processor, the support for
PH curves is necessarily limited--perhaps even to 3 control points.
Given more processing power and memory (e.g., BeagleBone Black) this library 
can traverse more complex PH curves having more control points.

Generally speaking, you want to generate PH curves on a fast, desktop/laptop,
but you can traverse them on more modest hardware.

### Installation: ArduinoMega2560
1. Download and install the [library](https://github.com/firepick1/ph5/releases):

	`https://github.com/firepick1/ph5/releases`

1. Unzip and copy the **ph5** folder under the **libraries** folder of your
Arduino workspace folder

	*your-personal-Arduino-workspace-folder*`/libraries/ph5`

1. Restart your Arduino IDE

1. To use the library, include the following line in your sketch:

	`#include <ph5.h>`


CHANGELOG.md
PH5: change log
===============

HEAD
----

v0.1.6
------
* performance improvements

v0.1.5
------
* 5 element tinyvector

v0.1
------
* Arduino library
* PH5TYPE is "float" by default
* sAccel bug



test/test.cpp
#include <string.h>
#include <iostream>
#include <fstream>
#include <sstream>
#include "FireLog.h"
#include "FireUtils.hpp"
#include "ph5.h"
#include "version.h"

using namespace std;
using namespace ph5;

void test_Complex() {
    cout << "TEST	: test_Complex()" << endl;

    Complex<PH5TYPE> c1(1, 2);
    ASSERTEQUAL(c1.Re(), 1);
    ASSERTEQUAL(c1.Im(), 2);

    Complex<PH5TYPE> c5(3, 4);
    ASSERTEQUAL(5, c5.modulus());

    Complex<PH5TYPE> c15 = c1 + c5;
    ASSERT(c15 == Complex<PH5TYPE>(4, 6));
    ASSERT(c1 + 1 == Complex<PH5TYPE>(2, 2));

    Complex<PH5TYPE> c13(1, 3);
    Complex<PH5TYPE> c24(2, 4);
    Complex<PH5TYPE> c1324 = c13 - c24;
    ASSERT((c13 - c24) == Complex<PH5TYPE>(-1, -1));
    ASSERT((c24 - c13) == Complex<PH5TYPE>(1, 1));
    ASSERT((c13 - 1) == Complex<PH5TYPE>(0, 3));

    ASSERT((c13 * c24) == Complex<PH5TYPE>(1 * 2 - 3 * 4, 1 * 4 + 3 * 2));
    ASSERT((c24 * c13) == Complex<PH5TYPE>(1 * 2 - 3 * 4, 1 * 4 + 3 * 2));
    ASSERT((c13 * 2) == Complex<PH5TYPE>(2, 6));

    ASSERT(c13.conj() == Complex<PH5TYPE>(1, -3));

    Complex<PH5TYPE> cr = c13.recip();
    PH5TYPE epsilon = 0.000001;
    ASSERTEQUALT(cr.Re(), 0.1, epsilon);
    ASSERTEQUALT(cr.Im(), -0.3, epsilon);
    Complex<PH5TYPE> crr = cr.recip();
    ASSERTEQUALT(crr.Re(), 1, epsilon);
    ASSERTEQUALT(crr.Im(), 3, epsilon);

    Complex<PH5TYPE> c13div24 = c13 / c24;
    ASSERTEQUALT(c13div24.Re(), 0.7, epsilon);
    ASSERTEQUALT(c13div24.Im(), 0.1, epsilon);

    Complex<PH5TYPE> sum1;
    sum1.add(c13);
    ASSERT(sum1 == Complex<PH5TYPE>(1, 3));
    sum1.add(c24);
    ASSERT(sum1 == Complex<PH5TYPE>(3, 7));

    ASSERT(Complex<PH5TYPE>(25).sqrt() == Complex<PH5TYPE>(5));
    ASSERT(Complex<PH5TYPE>(-1).sqrt() == Complex<PH5TYPE>(0, 1));
    ASSERT(Complex<PH5TYPE>(3, 4).sqrt() == Complex<PH5TYPE>(2, 1));

    ASSERTEQUALS("0", Complex<PH5TYPE>().stringify().c_str());
    ASSERTEQUALS("1", Complex<PH5TYPE>(1).stringify().c_str());
    ASSERTEQUALS("i", Complex<PH5TYPE>(0, 1).stringify().c_str());
    ASSERTEQUALS("-i", Complex<PH5TYPE>(0, -1).stringify().c_str());
    ASSERTEQUALS("2i", Complex<PH5TYPE>(0, 2).stringify().c_str());
    ASSERTEQUALS("-2i", Complex<PH5TYPE>(0, -2).stringify().c_str());
    ASSERTEQUALS("1+i", Complex<PH5TYPE>(1, 1).stringify().c_str());
    ASSERTEQUALS("1-i", Complex<PH5TYPE>(1, -1).stringify().c_str());
    ASSERTEQUALS("1+2i", Complex<PH5TYPE>(1, 2).stringify().c_str());
    ASSERTEQUALS("1-2i", Complex<PH5TYPE>(1, -2).stringify().c_str());
    ASSERTEQUALS("1.2-5.7i", Complex<PH5TYPE>(1.2345, -5.6789).stringify(1).c_str());
    ASSERTEQUALS("1.0+1.0i", Complex<PH5TYPE>(0.99, 0.99).stringify(1).c_str());
    ASSERTEQUALS("1.0-1.0i", Complex<PH5TYPE>(0.99, -0.99).stringify(1).c_str());
    ASSERTEQUALS("0.99-0.99i", Complex<PH5TYPE>(0.99, -0.99).stringify(2).c_str());

    c13.assertEqualT(Complex<PH5TYPE>(1.0009, 2.9991), 0.001);
    cout << "TEST	:   test_Complex() OK" << endl;
}

void test_Bernstein() {
    PH5TYPE  epsilon = 0.00001;
    cout << "TEST	: test_Bernstein()" << endl;
    ASSERTEQUALT(0, Bernstein5<PH5TYPE>(5, 0), epsilon);
    ASSERTEQUALT(0.03125, Bernstein5<PH5TYPE>(5, 0.5), epsilon);
    ASSERTEQUALT(0.15625, Bernstein5<PH5TYPE>(1, 0.5), epsilon);
    cout << "TEST	:   test_Bernstein() OK" << endl;
}

PH5Curve<PH5TYPE> ph_arc() {
    PHVECTOR<Complex<PH5TYPE> > q;
    PHVECTOR<Complex<PH5TYPE> > z;
    q.push_back(Complex<PH5TYPE>(-1, 1));
    q.push_back(Complex<PH5TYPE>(0, 2));
    q.push_back(Complex<PH5TYPE>(1, 1));
    z.push_back(Complex<PH5TYPE>());
    z.push_back(Complex<PH5TYPE>(1.124171968973597, 0.444771808762066));
    z.push_back(Complex<PH5TYPE>(1.124171968973597, -0.444771808762066));
    return PH5Curve<PH5TYPE>(z, q);
}

void test_PH5Curve() {
    cout << "TEST	: test_PH5Curve()" << endl;
    PH5Curve<PH5TYPE> ph(ph_arc());

    long msStart = millis();
    PH5TYPE epsilon = 0.00001;
    int16_t ITER = 10000;
    for (int16_t i = 0; i < ITER; i++) {
        Complex<PH5TYPE> c;
        c = ph.r(0);
        c.assertEqualT(Complex<PH5TYPE>(-1, 1), epsilon);
        c = ph.r(0.25);
        c.assertEqualT(Complex<PH5TYPE>(-0.598911, 1.75), epsilon);
        c = ph.r(0.5);
        c.assertEqualT(Complex<PH5TYPE>(0, 2), epsilon);
        c = ph.r(0.75);
        c.assertEqualT(Complex<PH5TYPE>(0.598911, 1.75), epsilon);
        c = ph.r(1.0);
        c.assertEqualT(Complex<PH5TYPE>(1, 1), epsilon);
    }
    long msElapsed = millis() - msStart;
    cout << "TEST	:   average execution time for r(): " << msElapsed / (ITER / 5000) << "us" << endl;

    ASSERTEQUALT(0.00000, ph.s(0.0), epsilon);
    ASSERTEQUALT(1.52753, ph.s(0.5), epsilon);
    ASSERTEQUALT(3.05505, ph.s(1.0), epsilon);

    ASSERTEQUALT(4.11010, ph.sigma(0.0), epsilon);
    ASSERTEQUALT(2.78074, ph.sigma(0.3), epsilon);
    ASSERTEQUALT(2.52753, ph.sigma(0.5), epsilon);
    ASSERTEQUALT(2.78074, ph.sigma(0.7), epsilon);
    ASSERTEQUALT(4.11010, ph.sigma(1.0), epsilon);

    epsilon = 0.001;
    ph.rprime(0.00).assertEqualT(Complex<PH5TYPE>(0.945, 4.000), epsilon);
    ph.rprime(0.25).assertEqualT(Complex<PH5TYPE>(2.132, 2.000), epsilon);
    ph.rprime(0.50).assertEqualT(Complex<PH5TYPE>(2.528, 0.000), epsilon);
    ph.rprime(0.75).assertEqualT(Complex<PH5TYPE>(2.132, -2.000), epsilon);
    ph.rprime(1.00).assertEqualT(Complex<PH5TYPE>(0.945, -4.000), epsilon);

    cout << "TEST	:   test_PH5Curve() OK " << endl;
}

void test_PHFeed() {
    cout << "TEST	: test_PH5Feed()" << endl;
    PH5Curve<PH5TYPE> ph(ph_arc());
    PHFeed<PH5TYPE> phf(ph, 100, 0.01);
    PH5TYPE E = 0;
    PH5TYPE epsilon = 0.001;
    ASSERTEQUALT(0, ph.s(E), epsilon);
    ph.r(E).assertEqualT(Complex<PH5TYPE>(-1.000, 1.000), epsilon);
    E = phf.Ekt(E, 0.1);
    ASSERTEQUALT(0.039, ph.s(E), epsilon);
    ph.r(E).assertEqualT(Complex<PH5TYPE>(-0.990, 1.038), 0.001);
    E = phf.Ekt(E, 0.2);
    ASSERTEQUALT(0.313, ph.s(E), epsilon);
    ph.r(E).assertEqualT(Complex<PH5TYPE>(-0.904, 1.298), 0.001);
    E = phf.Ekt(E, 0.3);
    ASSERTEQUALT(0.716, ph.s(E), epsilon);
    ph.r(E).assertEqualT(Complex<PH5TYPE>(-0.699, 1.643), 0.001);
    E = phf.Ekt(E, 0.4);
    ASSERTEQUALT(1.122, ph.s(E), epsilon);
    ph.r(E).assertEqualT(Complex<PH5TYPE>(-0.389, 1.901), 0.001);
    E = phf.Ekt(E, 0.5);
    ASSERTEQUALT(1.527, ph.s(E), epsilon);
    ph.r(E).assertEqualT(Complex<PH5TYPE>(-0.000, 2.000), 0.001);
    E = phf.Ekt(E, 0.6);
    ASSERTEQUALT(1.933, ph.s(E), epsilon);
    ph.r(E).assertEqualT(Complex<PH5TYPE>(0.389, 1.901), 0.001);
    E = phf.Ekt(E, 0.7);
    ASSERTEQUALT(2.338, ph.s(E), epsilon);
    ph.r(E).assertEqualT(Complex<PH5TYPE>(0.699, 1.643), 0.001);
    E = phf.Ekt(E, 0.8);
    ASSERTEQUALT(2.741, ph.s(E), epsilon);
    ph.r(E).assertEqualT(Complex<PH5TYPE>(0.904, 1.298), 0.001);
    E = phf.Ekt(E, 0.9);
    ASSERTEQUALT(3.015, ph.s(E), epsilon);
    ph.r(E).assertEqualT(Complex<PH5TYPE>(0.990, 1.038), 0.001);
    E = phf.Ekt(E, 1.0);
    ASSERTEQUALT(3.055, ph.s(E), epsilon);
    ph.r(E).assertEqualT(Complex<PH5TYPE>(1.000, 1.000), 0.001);

    cout << "TEST	:   test_PH5Feed() OK " << endl;
}

void test_SamplePHCurve() {
    cout << "TEST	: test_SamplePHCurve()" << endl;

    // Create a PHCurve interpolator from the PH curve coefficients
    // for an arc connecting the three points {(-1,1),(0,2),(1,1)}
    PH5Curve<PH5TYPE> ph(ph_arc());

    PH5TYPE vMax = 100; // maximum velocity (mm/s)
    PH5TYPE tvMax = 0.01; // time to achieve maximum velocity (seconds)
    PH5TYPE vIn = 0; // initial velocity (mm/s)
    PH5TYPE vCruise = vMax; // cruising velocity (mm/s)
    PH5TYPE vOut = 0; // final velocity (mm/s)

    // Create a quintic feedrate for traversing the above curve smoothly
    PHFeed<PH5TYPE> phf(ph, vMax, tvMax, vIn, vOut, vCruise);

    int16_t N = 100; // number of points to interpolate
    PH5TYPE E = 0; // interpolation normalized parametric state [0,1]

    // Generate a set of points using PH feed rate along PH curve
    cout << "X,Y" << endl;
    for (PH5TYPE tPoint = 0; tPoint <= N; tPoint++) {
        E = phf.Ekt(E, tPoint / N);
        Complex<PH5TYPE> point = ph.r(E);
        cout << point.Re() << "," << point.Im() << endl;
    }

    cout << "TEST	:   test_SamplePHCurve() OK " << endl;
}

void test_size() {
	cout << "sizeof(Complex):" << sizeof(Complex<PH5TYPE>) << endl;
	cout << "sizeof(PH5Curve):" << sizeof(PH5Curve<PH5TYPE>) << endl;
	cout << "sizeof(PHFeed):" << sizeof(PHFeed<PH5TYPE>) << endl;
}

void test_performance() {
    PH5Curve<PH5TYPE> ph(ph_arc());
    PHFeed<PH5TYPE> phf(ph, 100, 0.01);
    int16_t N = 100; // number of points to interpolate
    PH5TYPE E = 0; // interpolation normalized parametric state [0,1]
	int iterations = 1000000;

	int32_t msStart;
	int32_t msElapsed;

	msStart = millis();
	PH5TYPE s;
	PH5TYPE Ekr = 0.35; // arbitrary
	for (int iter=0; iter<iterations; iter++) {
		s = ph.s(Ekr); // 269-330ms baseline;  268-330 improved
	}
	msElapsed = millis() - msStart;
	cout << "testPerformance ph.s():" << s << " " << msElapsed << "ms"<< endl;

	msStart = millis();
	PH5TYPE sigma;
	for (int iter=0; iter<iterations; iter++) {
		sigma = ph.sigma(Ekr); // 309-330ms baseline; 205-213 improved
	}
	msElapsed = millis() - msStart;
	cout << "testPerformance ph.sigma():" << sigma << " " << msElapsed << "ms"<< endl;

	msStart = millis();
	Complex<PH5TYPE> r;
	for (int iter=0; iter<iterations; iter++) {
		r = ph.r(Ekr); // 196-210ms baseline
	}
	msElapsed = millis() - msStart;
	cout << "testPerformance ph.r():" << msElapsed << "ms"<< endl;

	msStart = millis();
	PH5TYPE tau = 0.81;
	PH5TYPE Ftau;
	for (int iter=0; iter<iterations; iter++) {
		Ftau = phf.Ft(tau); // 185-205ms baseline
	}
	msElapsed = millis() - msStart;
	cout << "testPerformance phf.Ft():" << msElapsed << "ms"<< endl;

	msStart = millis();
	for (int iter=0; iter<1000; iter++) {
		for (PH5TYPE tPoint = 0; tPoint <= N; tPoint++) {
			E = phf.Ekt(E, tPoint / N);
			Complex<PH5TYPE> point = ph.r(E);
		}
	}
	msElapsed = millis() - msStart; // 231-240ms improved
	cout << "testPerformance phf.Ekt() + ph.r(E):" << msElapsed << endl;
}

int main(int argc, char *argv[]) {
    LOGINFO3("INFO	: ph5 test v%d.%d.%d", VERSION_MAJOR, VERSION_MINOR, VERSION_PATCH);
    firelog_level(FIRELOG_TRACE);

    test_Complex();
    test_Bernstein();
    test_PH5Curve();
    test_PHFeed();
    test_SamplePHCurve();
	test_size();
	test_performance();

    cout << "TEST	: END OF TEST main()" << endl;
}


ph5.cpp
#include <string.h>
#include <iostream>
#include <fstream>
#include <sstream>
#include <math.h>
#include "version.h"
#include "ph5.h"

using namespace std;
using namespace ph5;

static void help() {
    cout << "ph5 Pythagorean Hodograph Quintic curve library" << VERSION_MAJOR << "." << VERSION_MINOR << "." << VERSION_PATCH << endl;
    cout << "Copyright 2015, Karl Lew" << endl;
    cout << "https://github.com/ph5/README.md" << endl;
}

bool parseArgs(int argc, char *argv[]) {
    firelog_level(FIRELOG_INFO);

    if (argc <= 1) {
        return false;
    }

    for (int16_t i = 1; i < argc; i++) {
        if (argv[i][0] == 0) {
            // empty argument
        } else if (strcmp("-version", argv[i]) == 0) {
            cout << "{\"version\":\"" << VERSION_MAJOR << "." << VERSION_MINOR << "." << VERSION_PATCH << "\"}" << endl;
            return false;
        } else if (strcmp("-warn", argv[i]) == 0) {
            firelog_level(FIRELOG_WARN);
        } else if (strcmp("-error", argv[i]) == 0) {
            firelog_level(FIRELOG_ERROR);
        } else if (strcmp("-info", argv[i]) == 0) {
            firelog_level(FIRELOG_INFO);
        } else if (strcmp("-debug", argv[i]) == 0) {
            firelog_level(FIRELOG_DEBUG);
        } else if (strcmp("-trace", argv[i]) == 0) {
            firelog_level(FIRELOG_TRACE);
        } else {
            LOGERROR1("unknown ph5 argument: '%s'", argv[i]);
            return false;
        }
    }
    return true;
}

int main(int argc, char *argv[]) {
	cout << "PH5 hello" << endl;
    bool argsOk = parseArgs(argc, argv);
    if (!argsOk) {
        help();
		return -1;
    }

	/*
    Mat image;
    if (imagePath) {
        LOGTRACE1("Reading image: %s", imagePath);
        image = imread(imagePath);
        if (!image.data) {
            LOGERROR1("main() imread(%s) failed", imagePath);
            exit(-1);
        }
    } else {
        LOGDEBUG("No image specified.");
    }

    switch (uimode) {
    case UI_STILL:
        uiStill(pipelinePath.c_str(), image, argMap, isTime, jsonIndent);
        break;
    case UI_VIDEO:
        uiVideo(pipelinePath.c_str(), argMap);
        break;
    default:
        LOGERROR("Unknown UI mode");
        exit(-1);
    }

    if (outputPath) {
        if (!imwrite(outputPath, image)) {
            LOGERROR1("Could not write image to: %s", outputPath);
            exit(-1);
        }
        LOGTRACE1("Image written to: %s", outputPath);
    }

*/
    return 0;
}


ph5.hpp
#ifndef PH5_HPP
#define PH5_HPP

#include <vector>
#include <map>
#ifdef _MSC_VER
#include "winjunk.hpp"
#else
#define CLASS_DECLSPEC
#endif
#include <math.h>

using namespace std;

#if __amd64__ || __x86_64__ || _WIN64 || _M_X64
#define PH5_64_BIT
#define PH5_PLATFORM_BITS 64
#else
#define PH5_32_BIT
#define PH5_PLATFORM_BITS 32
#endif

#ifndef PH5TYPE 
#define PH5TYPE float
#endif

namespace ph5 {

template <class T>
class CLASS_DECLSPEC Complex {
	private:
		T re;
		T im;

	public:
		Complex(T re=0, T im=0);
	public:
		inline T Re() { return this->re; }
	public:
		inline T Im() { return this->im; }
	public: 
		T modulus() { return ::sqrt(re*re + im*im); }
	public:
		inline Complex<T> operator+(Complex<T> that) { 
			return Complex(re+that.re, im+that.im);
		}
	public:
		inline Complex<T> operator+(T k) { 
			return Complex(re+k, im);
		}
	public:
		inline Complex<T> operator-(Complex<T> that) { 
			return Complex(re-that.re, im-that.im);
		}
	public:
		inline Complex<T> operator-(T k) { 
			return Complex(re-k, im);
		}
	public:
		inline bool operator==(Complex<T> that) {
			return re == that.re && im == that.im;
		}
	public:
		inline Complex<T> operator*(Complex<T> that) {
			return Complex(re*that.re-im*that.im,re*that.im+im*that.re);
		}
	public:
		inline Complex<T> operator*(T that) {
			return Complex(re*that,im*that);
		}
	public:
		inline Complex<T> conj() {
			return Complex(re, -im);
		}
	public:
		inline Complex<T> recip() {
			T denom = re*re + im*im;
			return Complex(re/denom, -im/denom);
		}
	public:
		inline Complex<T> operator/ (Complex<T> c2) {
			T denom = c2.re*c2.re + c2.im*c2.im;
			return Complex(
				(re*c2.re + im*c2.im)/denom,
				(im*c2.re - re*c2.im)/denom
			);
		}
	public:
		inline Complex<T> operator/ (T k) {
			return Complex(re/k, im/k);
		}
	public:
		inline void add(Complex<T> that) {
			re += that.re;
			im += that.im;
		}
	public:
		inline Complex<T> sqrt() {
			T m = modulus();
			T p = ::sqrt((m+re)/2);
			T q = ::sqrt((m-re)/2);
			if (im >= 0) {
				return Complex(p,q);
			} else {
				return Complex(p,-q);
			}
		}
	public:
		string stringify(int nPlaces=0);
	public:
		bool assertEqualT(Complex<T> that, double tolerance=0.0000001);
};

namespace ph5 {
template<class T>
inline Complex<T> operator*(T k, Complex<T> c) {
	return Complex<T>(k*c.re, k*c.im);
}
}

template<class T>
class CLASS_DECLSPEC PH5Curve {
	private:
		vector<Complex<T> > z;
		vector<Complex<T> > q;
		vector<Complex<T> > wi0;
		vector<Complex<T> > wi1;
		vector<Complex<T> > wi2;
		vector<Complex<T> > pi0;
		vector<Complex<T> > pi1;
		vector<Complex<T> > pi2;
		vector<Complex<T> > pi3;
		vector<Complex<T> > pi4;
		vector<Complex<T> > pi5;
		vector<T> sigmai0;
		vector<T> sigmai1;
		vector<T> sigmai2;
		vector<T> sigmai3;
		vector<T> sigmai4;
		int N;

	protected:
		Complex<T> calc_wij(int i, int j);
		Complex<T> pik(int i, int k);
		Complex<T> rit(int i, T p);
		T sit(int i, T p);
		T sik(int i, int j);
		T sigmaij(int i, int j);
		Complex<T> ritprime(int i, T p);

    public:
		PH5Curve(vector<Complex<T> > phz, vector<Complex<T> > phq);
		Complex<T> r(T p);
		T s(T p);
		T sigma(T p);
		Complex<T> rprime(T p);
};

template<class T>
class CLASS_DECLSPEC PHFeed {
	protected:
		PH5Curve<T> &ph;	// REQUIRED: curve to traverse
		T vMax;				// OPTION: maximum velocity
		T tvMax;			// OPTION: time from rest to maximum velocity
		T vIn;				// OPTION: PH curve entry velocity
		T vCruise;			// OPTION: PH curve constant cruise velocity
		T vOut;				// OPTION: PH curve exit velocity
		T S;				// curve arc length
		T sAccel;			// acceleration arc length distance
		T sCruise;			// cruise arc length distance
		T sDecel;			// deceleration arc length distance
		T tAccel;			// acceleration time
		T tCruise;			// cruise time
		T tDecel;			// deceleration time
		T tS;				// total traversal time
		T epsilon;			// Newton-Raphson convergence limit
		int iterations;		// maximum Newton-Raphson iterations 
		T tauCruise;
		T tauDecel;
		T Faccel[7];
		T Fcruise[7];
		T Fdecel[7];
		inline T Vaccel(int k) { return k < 3 ? vIn : vCruise; };
		inline T Vdecel(int k) { return k < 3 ? vCruise : vOut; };
		T Fk(T vIn, T vOut, int k);
		inline T Vk(T vIn, T vOut, int k) { return k<3 ? vIn : vOut; }

	public:
		PHFeed(PH5Curve<T> &ph5, T vMax=200, T tvMax=0.1, T vIn=0, T vCruise=200, T vOut=0);
	public:
		T F(T tau);
	public:
		T Ekt(T Ekprev, T tau);

	public:
		inline T sigma(T E) { return ph.sigma(E); }
		inline T s(T E) { return ph.s(E); }
		inline Complex<T> r(T E) { return ph.r(E); }
		inline T get_tS() { return tS; };
		inline T get_tAccel() { return tAccel; };
		inline T get_tCruise() { return tCruise; };
		inline T get_tDecel() { return tDecel; };
		inline T get_sAccel() { return sAccel; };
		inline T get_sCruise() { return sCruise; };
		inline T get_sDecel() { return sDecel; };
		inline T get_S() { return S; };
		inline T get_vIn() { return vIn; };
		inline T get_vCruise() { return vCruise; };
		inline T get_vOut() { return vOut; };
		inline T get_vMax() { return vMax; };
		inline T get_tvMax() { return tvMax; };
		
};

extern int choose5[6];
extern int choose6[7];

template <class T>
T Bernstein5(int k, T p) {
	T result = choose5[k];
	T p1 = 1 - p;
	for (int i=0; i<5-k; i++) {
		result *= p1;
	}
	for (int i=0; i<k; i++) {
		result *= p;
	}
	return result;
}

template <class T>
T Bernstein6(int k, T p) {
	T result = choose6[k];
	T p1 = 1 - p;
	for (int i=0; i<6-k; i++) {
		result *= p1;
	}
	for (int i=0; i<k; i++) {
		result *= p;
	}
	return result;
}

} // namespace ph5

#endif


Complex.cpp
#ifdef CMAKEx
#include <string.h>
#include <iostream>
#include <fstream>
#include <sstream>
#endif
#include <math.h>
#include "version.h"
#include "ph5.h"

using namespace std;
using namespace ph5;

template<class T>
Complex<T>::Complex(T re, T im) 
	: re(re), im(im) {
}

template<class T>
bool Complex<T>::assertEqualT(Complex<T> that, double tolerance) {
	ASSERTEQUALT(that.re, re, tolerance);
	ASSERTEQUALT(that.im, im, tolerance);
}

#ifdef CMAKE
template<class T>
string Complex<T>::stringify(int16_t nPlaces) {
	string s;
	char fmt[100];
	snprintf(fmt, sizeof(fmt), "%%.%dlf", nPlaces);
	char sre[100];
	snprintf(sre, sizeof(sre), fmt, (double) re);
	char sim[100];
	snprintf(sim, sizeof(sim), fmt, (double) im);
	if (im) {
		if (im < 0) {
			if (re != 0) {
				s.append(sre);
			}
			if (im == -1) {
				s.append("-");
			} else {
				s.append(sim);
			}
		} else {
			if (re != 0) {
				s.append(sre);
				s.append("+");
			}
			if (im != 1) {
				s.append(sim);
			}
		}
		s.append("i");
	} else {
		s.append(sre);
	}
	return s;
}
#endif

template class Complex<PH5TYPE>;


FireLog.cpp
#include "FireLog.h"
//#include "version.h"
#define VERSION_MAJOR 1
#define VERSION_MINOR 0
#define VERSION_PATCH 0

#include <iostream>

#include <errno.h>
#include <time.h>
#ifdef WIN32
#include <Winsock2.h>
#include <Windows.h>
#else
#define LOG_THREAD_ID
#include <sys/time.h>
#endif
#include <string.h>

#ifdef LOG_THREAD_ID
#ifndef _GNU_SOURCE
#define _GNU_SOURCE         /* See feature_test_macros(7) */
#endif
#include <unistd.h>
#include <sys/syscall.h>
static bool logTID = 1;
#else
static bool logTID = 0;
#endif

#define LOGMAX 255

using namespace std;

FILE *logFile = NULL;
int logLevel = FIRELOG_WARN;
static char lastMessage[5][LOGMAX+1];

static class Singleton {
	public: Singleton() {};
	public: ~Singleton() {
		if (logFile) {
			LOGINFO("==============CLOSING LOG FILE==============");
			fclose(logFile);
		}
	}
} singleton;

int firelog_init(const char *path, int level) {
    logLevel = level;
    logFile = fopen(path, "w");
    if (!logFile) {
        return errno;
    }
    char version[32];
    snprintf(version, sizeof(version), "%d.%d.%d", VERSION_MAJOR, VERSION_MINOR, VERSION_PATCH);
    LOGINFO2("FireLog %s version %s", path, version);
    firelog_lastMessageClear();
    return 0;
}

int firelog_destroy() {
    int rc = 0;
    if (logFile) {
        rc = fclose(logFile);
    }
    return rc;
}

/* Last message up to given level*/
const char * firelog_lastMessage(int level) {
    int i;
    for (i = 0; i<FIRELOG_TRACE; i++) {
        if (lastMessage[i]) {
            return lastMessage[i];
        }
    }
    return "";
}

void firelog_show_thread_id(int show) {
    logTID = show;
}

void firelog_lastMessageClear() {
    memset(lastMessage, 0, sizeof(lastMessage));
}

int firelog_level(int newLevel) {
    int oldLevel = logLevel;
    logLevel = newLevel;
    switch (newLevel) {
    case FIRELOG_ERROR:
        LOGDEBUG1("firelog_level(%s)", "FIRELOG_ERROR");
        break;
    case FIRELOG_WARN:
        LOGDEBUG1("firelog_level(%s)", "FIRELOG_WARN");
        break;
    case FIRELOG_INFO:
        LOGDEBUG1("firelog_level(%s)", "FIRELOG_INFO");
        break;
    case FIRELOG_DEBUG:
        LOGDEBUG1("firelog_level(%s)", "FIRELOG_DEBUG");
        break;
    case FIRELOG_TRACE:
        LOGDEBUG1("firelog_level(%s)", "FIRELOG_TRACE");
        break;
    default:
        LOGERROR1("firelog_level(unknown level %d)", newLevel);
        break;
    }
    return oldLevel;
}

void firelog(const char *msg, int level) {
#ifdef WIN32
    SYSTEMTIME st;
    GetSystemTime(&st);
    int now_hour = st.wHour;
    int now_min = st.wMinute;
    int now_sec = st.wSecond;
    int now_ms =  st.wMilliseconds;
#else
    timeval tp;
    gettimeofday(&tp, 0);
    time_t curtime = tp.tv_sec;
    struct tm *pLocalNow = localtime(&curtime);
    int now_hour = pLocalNow->tm_hour;
    int now_min = pLocalNow->tm_min;
    int now_sec = pLocalNow->tm_sec;
    int now_ms = tp.tv_usec/1000;
#endif
    int tid = 0;
#ifdef LOG_THREAD_ID
    tid = syscall(SYS_gettid);
#endif
    const char * levelStr = "?";

    switch (level) {
    case FIRELOG_ERROR:
        levelStr = " ERROR ";
        break;
    case FIRELOG_WARN:
        levelStr = " W ";
        break;
    case FIRELOG_INFO:
        levelStr = " I ";
        break;
    case FIRELOG_DEBUG:
        levelStr = " D ";
        break;
    case FIRELOG_TRACE:
        levelStr = " T ";
        break;
    }

    if (logTID) {
        snprintf(lastMessage[level], LOGMAX, "%02d:%02d:%02d.%03d %d %s %s",
                 now_hour, now_min, now_sec, now_ms, tid, levelStr, msg);
    } else {
        snprintf(lastMessage[level], LOGMAX, "%02d:%02d:%02d.%03d %s %s",
                 now_hour, now_min, now_sec, now_ms, levelStr, msg);
    }

    if (logFile) {
        fprintf(logFile, "%s", lastMessage[level]);
        fprintf(logFile, "\n");
        fflush(logFile);
    }
#ifdef __cplusplus
    else {
        cerr << lastMessage[level] << endl;
        cerr.flush();
    }
#endif
}
